File: src/builder/binary.cpp
#include "program/resolver.h"
#include "series/type/parallelopseries.h"

template <typename RealType> struct FuncMod { RealType operator()(RealType a, RealType b) const { return std::fmod(a, b); } };
template <typename RealType> struct FuncPow { RealType operator()(RealType a, RealType b) const { return std::pow(a, b); } };
template <typename RealType> struct FuncMinimum { RealType operator()(RealType a, RealType b) const { return std::fmin(a, b); } };
template <typename RealType> struct FuncMaximum { RealType operator()(RealType a, RealType b) const { return std::fmax(a, b); } };
template <typename RealType> struct FuncShrink { RealType operator()(RealType a, RealType b) const { return a > 0 ? (a <= b ? 0 : a - b) : a < 0 ? (a >= -b ? 0 : a + b) : a; } };
template <typename RealType> struct FuncClamp { RealType operator()(RealType a, RealType b) const { return std::isnan(a) ? a : std::fmax(-b, std::fmin(a, b)); } };
template <typename RealType> struct FuncNanTo { RealType operator()(RealType a, RealType b) const { return std::isnan(a) ? b : a; } };

template <template <typename> typename Operator>
void declBinaryOp(app::AppContext &context, program::Resolver &resolver, const char *funcName) {
    resolver.decl(funcName, [](float a, float b) -> float { return Operator<float>()(a, b); });
    resolver.decl(funcName, [](float a, double b) -> double { return Operator<double>()(a, b); });
    resolver.decl(funcName, [](double a, float b) -> double { return Operator<double>()(a, b); });
    resolver.decl(funcName, [](double a, double b) -> double { return Operator<double>()(a, b); });
    resolver.decl(funcName, [&context](series::DataSeries<float> *a, float b){
        auto op = [b](float a) {return Operator<float>()(a, b);};
        return new series::ParallelOpSeries<float, decltype(op), series::DataSeries<float>>(context, op, *a);
    });
    resolver.decl(funcName, [&context](series::DataSeries<double> *a, double b){
        auto op = [b](double a) {return Operator<double>()(a, b);};
        return new series::ParallelOpSeries<double, decltype(op), series::DataSeries<double>>(context, op, *a);
    });
    resolver.decl(funcName, [&context](float a, series::DataSeries<float> *b){
        auto op = [a](float b) {return Operator<float>()(a, b);};
        return new series::ParallelOpSeries<float, decltype(op), series::DataSeries<float>>(context, op, *b);
    });
    resolver.decl(funcName, [&context](double a, series::DataSeries<double> *b){
        auto op = [a](double b) {return Operator<double>()(a, b);};
        return new series::ParallelOpSeries<double, decltype(op), series::DataSeries<double>>(context, op, *b);
    });
    resolver.decl(funcName, [&context](series::DataSeries<float> *a, series::DataSeries<float> *b){
        return new series::ParallelOpSeries<float, Operator<float>, series::DataSeries<float>, series::DataSeries<float>>(context, Operator<float>(), *a, *b);
    });
    resolver.decl(funcName, [&context](series::DataSeries<double> *a, series::DataSeries<double> *b){
        return new series::ParallelOpSeries<double, Operator<double>, series::DataSeries<double>, series::DataSeries<double>>(context, Operator<double>(), *a, *b);
    });
}

static int _ = program::Resolver::registerBuilder([](app::AppContext &context, program::Resolver &resolver) {
    declBinaryOp<std::plus>(context, resolver, "add");
    declBinaryOp<std::minus>(context, resolver, "sub");
    declBinaryOp<std::multiplies>(context, resolver, "mul");
    declBinaryOp<std::divides>(context, resolver, "div");
    declBinaryOp<FuncMod>(context, resolver, "mod");
    declBinaryOp<FuncPow>(context, resolver, "pow");
    declBinaryOp<std::less>(context, resolver, "lt");
    declBinaryOp<std::less_equal>(context, resolver, "lte");
    declBinaryOp<std::greater>(context, resolver, "gt");
    declBinaryOp<std::greater_equal>(context, resolver, "gte");
    declBinaryOp<std::equal_to>(context, resolver, "eq");
    declBinaryOp<std::not_equal_to>(context, resolver, "neq");
    declBinaryOp<FuncMinimum>(context, resolver, "min");
    declBinaryOp<FuncMaximum>(context, resolver, "max");
    declBinaryOp<FuncShrink>(context, resolver, "shrink");
    declBinaryOp<FuncClamp>(context, resolver, "clamp");
    declBinaryOp<FuncNanTo>(context, resolver, "nan_to");
});

File: tests/binary.ts
import {
  add,
  d,
  div,
  gt,
  input,
  lt,
  max,
  min,
  mod,
  mul,
  shrink,
  sub,
} from '../../trader-exprs/modules/base.ts';

const r = d;

export default [
  {
    name: `Test add`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: add(r(input('x')), r(input('y'))),
    output: {
      0: { z: NaN },
      1: { z: 11 },
      2: { z: NaN },
      3: { z: -3 },
      4: { z: NaN },
      5: { z: 15 },
      6: { z: 7 },
      7: { z: 123 },
      8: { z: 14 },
      9: { z: 234 },
    },
  },
  {
    name: `Test sub`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: sub(r(input('x')), r(input('y'))),
    output: {
      0: { z: NaN },
      1: { z: -3 },
      2: { z: NaN },
      3: { z: 7 },
      4: { z: NaN },
      5: { z: -29 },
      6: { z: 5 },
      7: { z: 123 },
      8: { z: 0 },
      9: { z: -234 },
    },
  },
  {
    name: `Test mul`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: mul(r(input('x')), r(input('y'))),
    output: {
      0: { z: NaN },
      1: { z: 28 },
      2: { z: NaN },
      3: { z: -10 },
      4: { z: NaN },
      5: { z: -154 },
      6: { z: 6 },
      7: { z: 0 },
      8: { z: 49 },
      9: { z: 0 },
    },
  },
  {
    name: `Test div`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: div(r(input('x')), r(input('y'))),
    output: {
      0: { z: NaN },
      1: { z: 4 / 7 },
      2: { z: NaN },
      3: { z: -0.4 },
      4: { z: NaN },
      5: { z: -7 / 22 },
      6: { z: 6 },
      7: { z: NaN },
      8: { z: 1 },
      9: { z: 0 },
    },
  },
  {
    name: `Test mod`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: mod(r(input('x')), r(input('y'))),
    output: {
      0: { z: NaN },
      1: { z: 4 },
      2: { z: NaN },
      3: { z: 2 },
      4: { z: NaN },
      5: { z: -7 },
      6: { z: 0 },
      7: { z: NaN },
      8: { z: 0 },
      9: { z: 0 },
    },
  },
  {
    name: `Test lt`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: lt(r(input('x')), r(input('y'))),
    output: {
      0: { z: 0 },
      1: { z: 1 },
      2: { z: 0 },
      3: { z: 0 },
      4: { z: 0 },
      5: { z: 1 },
      6: { z: 0 },
      7: { z: 0 },
      8: { z: 0 },
      9: { z: 1 },
    },
  },
  {
    name: `Test gt`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: gt(r(input('x')), r(input('y'))),
    output: {
      0: { z: 0 },
      1: { z: 0 },
      2: { z: 0 },
      3: { z: 1 },
      4: { z: 0 },
      5: { z: 0 },
      6: { z: 1 },
      7: { z: 1 },
      8: { z: 0 },
      9: { z: 0 },
    },
  },
  {
    name: `Test min`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: min(r(input('x')), r(input('y'))),
    output: {
      0: { z: NaN },
      1: { z: 4 },
      2: { z: 8 },
      3: { z: -5 },
      4: { z: 6 },
      5: { z: -7 },
      6: { z: 1 },
      7: { z: 0 },
      8: { z: 7 },
      9: { z: 0 },
    },
  },
  {
    name: `Test max`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: max(r(input('x')), r(input('y'))),
    output: {
      0: { z: NaN },
      1: { z: 7 },
      2: { z: 8 },
      3: { z: 2 },
      4: { z: 6 },
      5: { z: 22 },
      6: { z: 6 },
      7: { z: 123 },
      8: { z: 7 },
      9: { z: 234 },
    },
  },
  {
    name: `Test shrink`,
    variant: 'test-csl2-6',
    input: {
      0: { x: NaN, y: NaN },
      1: { x: 4, y: 7 },
      2: { x: NaN, y: 8 },
      3: { x: 2, y: -5 },
      4: { x: 6, y: NaN },
      5: { x: -7, y: 22 },
      6: { x: 6, y: 1 },
      7: { x: 123, y: 0 },
      8: { x: 7, y: 7 },
      9: { x: 0, y: 234 },
    },
    program: shrink(r(input('x')), r(input('y'))),
    output: {
      0: { z: NaN },
      1: { z: 0 },
      2: { z: NaN },
      3: { z: 7 },
      4: { z: NaN },
      5: { z: 0 },
      6: { z: 5 },
      7: { z: 123 },
      8: { z: 0 },
      9: { z: 0 },
    },
  },
];
